CASMPET-5297 Add Keycloak kcadm user management application notes. (#1263)
Cray-HPE/docs-csm: operations/security_and_authentication/Keycloak_User_Management_with_Kcadm.md

# Keycloak User Management with kcadm.sh

> 

Reviewer: General notes:
- All command prompts in the code blocks should be ncn-m001
- Only one space is needed at the beginning of each sentence
- Add a blank line between lines of text and code blocks

> The Cray CLI requires a valid Keycloak JWT token.  If the token is not present it must be obtained by supplying valid user credentials at the time of CLI initialization.  During the installation or upgrade of a Shasta system the Keycloak user LDAP federation state or credentials may not be known or the external LDAP system for which the user may have been federated may not be available.  These scenarios make it impossible to initialize the Cray CLI with a previously valid Keycloak user or LDAP federated Keycloak user.  If for whatever reason the Keycloak UI is not available to the admin to verify the state of the user, using Keycloak's kcadm.sh utility from the command line will be required.

Reviewer: `kcadm.sh` should be in backticks

> The kcadm.sh utility is included with the Shasta Keycloak container image and can be used as described below to perform troubleshooting operations that can verify the existence of a user, determine if the user is federated, change the user credential (for a keycloak-local account only) or create a new local Keycloak user that is compatible (has the correct role mappings) with the Cray CLI.  All the operations described below must be run as root from the ncn-m001 node.

Reviewer: ```suggestion
The `kcadm.sh` utility is included with the Shasta Keycloak container image and can be used as described below to perform troubleshooting operations that can verify the existence of a user, determine if the user is federated, change the user credential (for a keycloak-local account only), or create a new local Keycloak user that is compatible (has the correct role mappings) with the Cray CLI.  All the operations described below must be run as root from the `ncn-m001` node.
```

> Usage of the bash function(s) provided below will require addition into the Linux environment.  This can be accomplished by pasting the function(s) into your shell or pasting the function(s) into a file and then sourcing the file before running the indicated function(s).

Reviewer: ```suggestion
Usage of the bash function(s) provided below will require making additions to the Linux environment.  This can be accomplished by pasting the function(s) into your shell or pasting the function(s) into a file and then sourcing the file before running the indicated function(s).
```

> ## Checking the Status of a Keycloak User

Reviewer: ```suggestion
## Check the Status of a Keycloak User
```

> Use this procedure to determine if a user both exists in Keycloak and if the user is LDAP federated or not.  If the user is LDAP federated and the credential is not known, update the user credential in LDAP.  It will not be possible to use a federated user account to initialize the Cray CLI if LDAP is not available.  In this case you may create a local Keycloak user for use with the Cray CLI by following the procedure below for creating a local Keycloak Cray CLI user.  The credential for a local account may be updated if needed by following the procedure below for updating a local Keycloak account credential.

Reviewer: ```suggestion
Use this procedure to determine if a user both exists in Keycloak and if the user is LDAP federated or not.  If the user is LDAP federated and the credential is not known, update the user credential in LDAP.  It will not be possible to use a federated user account to initialize the Cray CLI if LDAP is not available.  In this case, create a local Keycloak user for use with the Cray CLI by following the procedure below for creating a local Keycloak Cray CLI user.  The credential for a local account may be updated if needed by following the procedure below for updating a local Keycloak account credential.
```

Example when the user account is a local Keycloak user (no 'federationLink' attribute):
```bash
ncn-m001:~ # kc-find-user localcli
Looking for the Keycloak user: localcli
Logging into http://localhost:8080/keycloak as user admin of realm master
[ {
  "id" : "b49abdcc-a314-4355-89d7-44f4d8d33ab8",
  "createdTimestamp" : 1649193410108,
  "username" : "localcli",
  "enabled" : true,
  "totp" : false,
  "emailVerified" : false,
  "disableableCredentialTypes" : [ ],
  "requiredActions" : [ ],
  "notBefore" : 0,
  "access" : {
    "manageGroupMembership" : true,
    "view" : true,
    "mapRoles" : true,
    "impersonate" : true,
    "manage" : true
  }
} ]
```

If the user does not exist and it is necessary to create a local Keycloak user for use with the Cray CLI, you may follow the procedure below for creating a local Keycloak Cray CLI user.

Example when the user account is not known to Keycloak:
```bash
ncn-m001:~ # kc-find-user nouser
Looking for the Keycloak user: nouser
Logging into http://localhost:8080/keycloak as user admin of realm master

[ ]ncn-m001:~ #
```

> ## Updating a Local Keycloak User Credential

Reviewer: ```suggestion
## Update a Local Keycloak User Credential
```

}
```

Example of updating the credential with an overridden password:
```bash
ncn-m001:~ # KC_USER_NEWPASSWD=wax0nwax0ff && kc-set-local-user-password localcli
Resetting password for the Keycloak user: localcli
Logging into http://localhost:8080/keycloak as user admin of realm master
```

If the user can not be found by username, the account credential will not be updated as in this example:
```bash
ncn-m001:~ # kc-set-local-user-password localc
Resetting password for the Keycloak user: localc
Logging into http://localhost:8080/keycloak as user admin of realm master
User not found for username: localc
command terminated with exit code 1
```

If the user is LDAP federated, the account credential will not be updated as in this example:
```bash
ncn-m001:~ # KC_USER_NEWPASSWD=wax0nwax0ff && kc-set-local-user-password vers
Resetting password for the Keycloak user: vers
Logging into http://localhost:8080/keycloak as user admin of realm master
null [Can't reset password as account is read only]
command terminated with exit code 1
```

> ## Creating a Local Keycloak Cray CLI User

Reviewer: ```suggestion
## Create a Local Keycloak Cray CLI User
```

    ./opt/jboss/keycloak/bin/kcadm.sh add-roles -r shasta --uusername "$1" --cclientid cray --rolename admin --no-config --server http://localhost:8080/keycloak --realm master --user admin --client admin-cli --password $MAA && \
    echo 'Adding the Shasta client admin role to the user: "$1"' && \
    ./opt/jboss/keycloak/bin/kcadm.sh add-roles -r shasta --uusername "$1" --cclientid shasta --rolename admin --no-config --server http://localhost:8080/keycloak --realm master --user admin --client admin-cli --password $MAA"
}
```

Example of creating a user:

```bash
ncn-m001:~ # kc-create-local-cli-user localcli
Creating the Keycloak user: localcli
Logging into http://localhost:8080/keycloak as user admin of realm master
Created new user with id 'b49abdcc-a314-4355-89d7-44f4d8d33ab8'
Resetting password for the Keycloak user: localcli
Logging into http://localhost:8080/keycloak as user admin of realm master
Adding the Cray client admin role to the user: localcli
Logging into http://localhost:8080/keycloak as user admin of realm master
Adding the Shasta client admin role to the user: localcli
Logging into http://localhost:8080/keycloak as user admin of realm master
```

> You can optionally search for the user using kc-find-user to check that the user has been created.  It will also be possible to initialize the Cray CLI using this new user and credential.  Refer to the Cray CLI documentation for details.

Reviewer: `kc-find-user` should be in backticks
